### PR TITLE
BBBIA-32.1

### PR DESCRIPTION
**Pull request by DevMile**

### CODE REVIEW

File: BigBlueButton-RLP/Constants/Script.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
+//
+//  Script.swift
+//  BigBlueButton-RLP
+//
+//  Created by Milan Bojic on 20.12.21..
+//
+
+import Foundation
+
+struct Script {
```
Comment (kwlz) [inline BigBlueButton-RLP/Constants/Script.swift:10]: Good idea to extract this.

File: BigBlueButton-RLP/Constants/Script.swift [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+             Conclusion: Adding javascript listener for mute button states is not the best approach.
+    */
+    /* TO DO: Decide what is the best way to observe the microphone audio enable/disable events in webView BBB html client.
+              Propose to receive microphone enabled/disabled events via websocket because we need to regulate
+              webRTC audio transfer state in Broadcast Upload Extension where socket and webRTC are instantiated.
+    */
+    static let muteButtonMessage = "muteButtonMessage"
+    static let muteButtonListener =
+                """
+                document.getElementsByClassName('lg--Q7ufB buttonWrapper--x8uow muteToggle--LY4Tr')[0].addEventListener('click', (event) => {
```
Comment (kwlz) [inline BigBlueButton-RLP/Constants/Script.swift:54]: This looks like a very volatile ID. Is there a better way to identify the element?

File: BigBlueButton-RLP/Constants/Script.swift (hunk 1 above)
Comment (kwlz) [inline BigBlueButton-RLP/Constants/Script.swift:10]: Could/should this be an `enum`? https://stackoverflow.com/questions/38585344/swift-constants-struct-or-enum

File: BigBlueButton-RLP/Scenes/WebViewController.swift [hunk 3]
```diff
@@ -91,15 +97,25 @@ class WebViewController: UIViewController, WKUIDelegate {
...
         // Fire event to execute javascript
-        webView.evaluateJavaScript(Constants.fireJSEvent) { (_, error) in
+        webView.evaluateJavaScript(script) { (_, error) in
             if error != nil {
                 print("⚡️☠️ Error executing injected javascript script ☠️⚡️")
             }
         }
     }
     
+    private func getJSessionId() {
```
Comment (kwlz) [inline BigBlueButton-RLP/Scenes/WebViewController.swift:109]: It's unusual to have a function start with `get` and not have it return anything. What would be a better name?

File: BigBlueButton-RLP/Scenes/WebViewController.swift [hunk 4]
```diff
@@ -154,18 +171,25 @@ extension WebViewController: WKNavigationDelegate {
...
         guard let messageBody = message.body as? [String: Any] else { return }
-        let payload = Data((messageBody["payload"] as? String)!.utf8)
-        do {
-            let jsData = try decoder.decode(JavascriptData.self, from: payload)
-            isPayloadReceived = true
-            if let encodedData = try? encoder.encode(jsData) {
-                defaults?.set(encodedData, forKey: Constants.javascriptData)
+        switch message.name {
+        case Script.muteButtonMessage:
+            print(messageBody)
```
Comment (kwlz) [inline BigBlueButton-RLP/Scenes/WebViewController.swift:177]: Is this a temporary solution?

File: README.md [hunk 5]
```diff
@@ -1,2 +1,52 @@
 # BBB-iOS
+Extending BBB ios mobile application to support screen sharing.
```
Comment (kwlz) [inline README.md:2]: Capitalization of iOS.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,52 @@
...
+5. NOTE: Notice there is a grey recording button in a top left corner - it is just temporary button until final implementation is done with BBB developers.
+6. Click on a grey recording button and Broadcast Upload Extension will start, offering to start recording for BBB app. Click Start Recording.
+7. BBBBroadcast is the name of Broadcast Upload Extension which is receiving video frames and audio samples in its class SampleHandler. 
+8. Audio/Video samples are sent to BBB server through WebRTC.
+
+
+## Flow
+* When app is run and active, WebViewController will load BBB web page and user must sign in.
+* As a meeting host you create a meeting, allow microphone and camera usage.
+* Now everything function through webView and you can use it as it is.
```
Comment (kwlz) [inline README.md:25]: "functions"

File: README.md [hunk 7]
```diff
@@ -1,2 +1,52 @@
...
+* If you want to start screen sharing - click on a grey recording button in top left corner.
```
Comment (kwlz) [inline README.md:26]: "the grey recording button"

File: README.md [hunk 8]
```diff
@@ -1,2 +1,52 @@
...
+* NOTE: The grey recording button will be removed and BBB screen share button will be available in webView.
+* We use javascript injection to obtain data from webView loaded page and observe event of tapping on screen share button (when it become available).
```
Comment (kwlz) [inline README.md:28]: "becomes"

File: README.md [hunk 9]
```diff
@@ -1,2 +1,52 @@
...
+* When screen sharing is tapped, Broadcast Upload Extension will offer screen recording and will start its flow.
+* First websocket connection is made with BBB server in order to exchange and set WebRTC SDP offer and answer, as well as ICE candidates.
+* Then everything is handled from SampleHandler class of Broadcast Upload Extension.
+* WebRTCClient is a wrapper for WebRTC library and is main engine for sending audio/video samples to BBB server.
+
+
+## WebRTC:
+WebRTC is an open-source project (libjingle_peerConnection) maintained by google with high-level API implementations for both iOS and Android. 
+WebRTC api can be read from [here](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API).
```
Comment (kwlz) [inline README.md:37]: "API" capitalization

File: README.md [hunk 10]
```diff
@@ -1,2 +1,52 @@
...
+WebRTC Framework used in project has been build from [here](https://github.com/stasel/WebRTC)
```
Comment (kwlz) [inline README.md:38]: "built"

File: README.md [hunk 11]
```diff
@@ -1,2 +1,52 @@
...
+
+
+## References:
+* WebRTC website: https://webrtc.org/
+* WebRTC source code: https://webrtc.googlesource.com/src
+* WebRTC iOS compile guide: https://webrtc.github.io/webrtc-org/native-code/ios/
+* appear.in dev blog post: https://github.com/appearin/tech.appear.in/blob/master/source/_posts/Getting-started-with-WebRTC-on-iOS.md (it uses old WebRTC api but still very informative)
```
Comment (kwlz) [inline README.md:45]: "API" capitalization

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by kwlz** [comment]
Thanks for adding a README!

File: BigBlueButton-RLP/Constants/Script.swift (hunk 2 above)
Comment (DevMile) [reply BigBlueButton-RLP/Constants/Script.swift:54]: Right now there is no other way, but this approach is not good at all. The best would be to receive mute button events through websocket. That will be my proposal to Tiago. This is now temporary solution which is not complete.

File: BigBlueButton-RLP/Constants/Script.swift (hunk 1 above)
Comment (DevMile) [reply BigBlueButton-RLP/Constants/Script.swift:10]: Yes, good point, dont know why I didnt make it enum :)

File: BigBlueButton-RLP/Scenes/WebViewController.swift (hunk 3 above)
Comment (DevMile) [reply BigBlueButton-RLP/Scenes/WebViewController.swift:109]: Yes, we can name it saveSessionCookie() or storeSessionCookie()?

File: BigBlueButton-RLP/Scenes/WebViewController.swift (hunk 4 above)
Comment (DevMile) [reply BigBlueButton-RLP/Scenes/WebViewController.swift:177]: Yes, it's temporary. Wont be here.

File: BigBlueButton-RLP/Scenes/WebViewController.swift (hunk 3 above)
Comment (kwlz) [reply BigBlueButton-RLP/Scenes/WebViewController.swift:109]: Both fine with me. 👍

**Review by ghost — APPROVED** [review]

File: BigBlueButton-RLP/Constants/Script.swift (hunk 2 above)
Comment (ghost) [reply BigBlueButton-RLP/Constants/Script.swift:54]: Not sure I understand the web socket proposal. @anton-sarudko will investigate further.

Comment (DevMile) [reply BigBlueButton-RLP/Constants/Script.swift:54]: Hi @kwlz, please note that this PR is merged to main. Regarding mute button, we need to wait to session with Tiago, he said he knows how this should be implemented so right now whole this part of code could be removed. Let us wait with this until session with Tiago. I dont think @anton-sarudko can investigate this properly without Tiago, too.

**Pull request merged**